classes/db/Db.php

```diff
@@ -94,91 +99,195 @@
+    }
 
     /**
      * Closes database connection
      */
-    abstract public function disconnect();
+    public function disconnect()
+    {
+        $this->link->close();
+    }
 
     /**
      * Execute a query and get result resource
      *
      * @param string $sql
      * @return PDOStatement|mysqli_result|resource|bool
      */
-    abstract protected function _query($sql);
+     protected function _query($sql)
+     {
+         return $this->link->executeQuery($sql);
+     }
 
     /**
      * Get number of rows in a result
      *
-     * @param mixed $result
+     * @param \Doctrine\DBAL\Driver\Statement $result
      * @return int
      */
-    abstract protected function _numRows($result);
+    protected function _numRows($result)
+    {
+        return $result->rowCount();
+    }
 
     /**
      * Get the ID generated from the previous INSERT operation
      *
      * @return int|string
      */
-    abstract public function Insert_ID();
+    public function Insert_ID()
+    {
+        return $this->link->lastInsertId();
+    }
 
     /**
      * Get number of affected rows in previous database operation
      *
      * @return int
      */
-    abstract public function Affected_Rows();
+    public function Affected_Rows()
+    {
+        return $this->result->rowCount();
+    }
 
     /**
      * Get next row for a query which does not return an array
      *
      * @param PDOStatement|mysqli_result|resource|bool $result
      * @return array|object|false|null
      */
-    abstract public function nextRow($result = false);
+    public function nextRow($result = false)
+    {
+        if (!$result) {
+            $result = $this->result;
+        }
+
+        if (!is_object($result)) {
+            return false;
+        }
+
+        return $result->fetch();
+    }
 
     /**
      * Get all rows for a query which return an array
      *
      * @param PDOStatement|mysqli_result|resource|bool|null $result
      * @return array
      */
-    abstract protected function getAll($result = false);
+    protected function getAll($result = false)
+    {
+        if (!$result) {
+            $result = $this->result;
+        }
+
+        if (!is_object($result)) {
+            return false;
+        }
+
+        return $result->fetchAll();
+    }
 
     /**
      * Get database version
      *
      * @return string
      */
-    abstract public function getVersion();
+    public function getVersion()
+    {
+        // ToDo: Check if best answer
+        return $this->getValue('SELECT VERSION()');
+    }
 
     /**
      * Protect string against SQL injections
      *
      * @param string $str
      * @return string
      */
-    abstract public function _escape($str);
+    public function _escape($str)
+    {
+        $search = array("\\", "\0", "\n", "\r", "\x1a", "'", '"');
+        $replace = array("\\\\", "\\0", "\\n", "\\r", "\Z", "\'", '\"');
+        return str_replace($search, $replace, $str);    }
```

Comment:
It seems quote() is used for a specific variable before being inserted in a sql request. With _escape, we search in the  whole sql request .